### PR TITLE
tetration_user_annotations initial commit

### PR DESCRIPTION
**Pull request by techBeck03**
##### SUMMARY

Adding Cisco Tetration module for managing user annotations

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

tetration_user_annotations

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: .github/BOTMETA.yml [hunk 1]
```diff
@@ -1604,7 +1609,8 @@ macros:
   team_rhn: alikins barnabycourt FlossWare vritant
   team_scaleway: remyleone abarbare DenBeke jerome-quere kindermoumoute QuentinBrosse
   team_solaris: bcoca dagwieers danowar2k fishman jpdasma jasperla mator scathatheworm troy2914 xen0l
-  team_suse: evrardjp toabctl dcermak lrupp commel
+  team_suse: evrardjp toabctl
```
Comment (Qalthos) [inline .github/BOTMETA.yml]: Why did you change `team_suse` here?

File: lib/ansible/plugins/doc_fragments/tetration.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+#
+# (c) 2015, Peter Sprygada <psprygada@ansible.com>
```
Comment (Qalthos) [inline lib/ansible/plugins/doc_fragments/tetration.py]: I am almost entirely sure this is not what you meant to have here

File: lib/ansible/module_utils/network/tetration.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+            raise Exception('tetpyclient is required but does not appear '
+                            'to be installed.  It can be installed using the '
+                            'command `pip install tetpyclient`')
+        if len(provider.keys()) > 0:
+            for key in provider.keys():
+                if key not in TETRATION_PROVIDER_SPEC.keys():
+                    del provider[key]
+        for key, value in iteritems(TETRATION_PROVIDER_SPEC):
+            if key not in provider:
+                # apply default values from NIOS_PROVIDER_SPEC since we cannot just
```
Comment (Qalthos) [inline lib/ansible/module_utils/network/tetration.py]:
```suggestion
                # apply default values from TETRATION_PROVIDER_SPEC since we cannot just
```
Is this what you meant?

File: lib/ansible/plugins/doc_fragments/tetration.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      api_version:
+        description:
+          - Specifies the version of Tetration OpenAPI to use
+          - Value can also be specified using C(TETRATION_API_VERSION) environment
+            variable.
+        type: str
+        default: v1
+notes:
+  - "This module must be run locally, which can be achieved by specifying C(connection: local)."
+  - Please read the :ref:`tetration_guide` for more detailed information on how to use Tetration with Ansible.
```
Comment (Qalthos) [inline lib/ansible/plugins/doc_fragments/tetration.py:108]: I'm not seeing where you have added the platform guide for tetration

File: lib/ansible/module_utils/network/tetration.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+                # assume the provider values are coming from AnsibleModule
+                if 'default' in value:
+                    provider[key] = value['default']
+                # override any values with env variables unless they were
+                # explicitly set
+                env = ('TETRATION_%s' % key).upper()
```
Comment (Qalthos) [inline lib/ansible/module_utils/network/tetration.py]: Do not do this. If you want parameters to fallback to environment variables then you can do the following in your TETRATION_PROVIDER_SPEC

```
'server_endpoint': dict(type='str', required=True, aliases=['endpoint', 'host'], fallback=(env_fallback, ['TETRATION_SERVER_ENDPOINT'])),
```

File: lib/ansible/module_utils/network/tetration.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+            if k not in list(iterkeys(obj2)):
+                unknown_keys.append(k)
+        return unknown_keys
+
+    def clear_values(self, obj):
+        for k in list(iterkeys(obj)):
+            obj[k] = ''
+
+
+TETRATION_API_PROTOCOLS = [
```
Comment (Qalthos) [inline lib/ansible/module_utils/network/tetration.py:227]: Is this used somewhere that I'm not seeing?

**Review by Qalthos — CHANGES_REQUESTED** [review]

**Comment by techBeck03** [comment]
I've pushed the changes you suggested.  The *TETRATION_API_PROTOCOLS* variable
is used in other modules we plan to submit once this initial submission is
complete.

Thank you so much for your help!

File: lib/ansible/plugins/doc_fragments/tetration.py (hunk 4 above)
Comment (techBeck03) [reply lib/ansible/plugins/doc_fragments/tetration.py:108]: The tetration reference guide is hosted within the cluster so there is no URL for me to provide.

File: lib/ansible/module_utils/network/tetration.py (hunk 6 above)
Comment (techBeck03) [reply lib/ansible/module_utils/network/tetration.py:227]: The `TETRATION_API_PROTOCOLS` variable
is used in other modules we plan to submit once this initial submission is
complete.

**Comment by Akasurde** [comment]
Hi @techBeck03, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.network<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
